src/python/WMComponent/WorkflowUpdater/WorkflowUpdaterPoller.py

```diff
@@ -26,6 +32,160 @@
+    return bdict
+
+
+def rucioBlockLocations(rucio, blocks):
+    """
+    Return dictionary in form: {"path-to-json1": jdoc1, "path-to-json2": jdoc2}
+    :param rucio: rucio client
+    :param blocks: list of blocks
+    :return: directionary
+    """
+    # rucio returns: [{"name": blockName, "replica": list(set(rses))}, ...]
+    rucioList = rucio.getReplicaInfoForBlocks(block=blocks)
+    bdict = {}
+    for item in rucioList:
+        bdict[item['name']] = item['replica']
+    return bdict
+
+
+def checkChanges(rdict, bdict):
+    """
+    Check if there is actually any pileup changes
+    :param rdict: dict of blocks locations from rucio
+    :param bdict: dict of blocks locations from pileupconf.json sandbox file
+    :return: boolean
+    """
+    # compare block names
+    if len(rdict) != len(bdict):
+        return True
+
+    # compare lcoations
+    for key in rdict.keys():
+        if sorted(rdict[key]) != sorted(bdict[key]):
+            return True
+    return False
+
+
+def updateBlockInfo(jdoc, rdict, bdict):
+    """
+    Discard given pileup names from JSON sandbox and return new dict.
+    NOTE: this function may consume lots of memory due to jdoc structure which
+    we need to parse and create new dict since we can't discard keys in place 
+    of jdoc internal dictionary, i.e.
+
+    jdoc structure is
+    {"<type: mc>": {"<block>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []},
+                    "<block>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []}, ...}
+    and we need to discard block names. They are located in internal dictionary
+    and in place pop-up of jdoc is not allowed since it will change size of internal dictionary.
+    Therefore, to overcome this obstacles we need to copy relevant pileupName key:values
+    into new dict structure and return new dict. Such operation will mostly double the
+    memory size.
+
+    :param jdoc: JSON sandbox dictionary
+    :param rdict: dict of blocks locations from rucio
+    :param bdict: dict of blocks locations from pileupconf.json sandbox file
+    :return: newly constructed dict
+    """
+    newDict = {}
+    for key, rec in jdoc.items():
+        newRec = {}
+        for blk in rec.keys():
+            # check if block present in the new container
+            if blk not in bdict.keys():
+                continue
+            loc = rec[blk]['PhEDExNodeNames']
+            rses = rdict.get(blk, bdict[blk])
+            if loc != rses:
+                rec[blk]['PhEDExNodeNames'] = rses
+            newRec[blk] = rec[blk]
+        newDict[key] = newRec
+    return newDict
+
+
+def writePileupJson(tfile, jdict, dest=None):
+    """
+    Write pileup JSON sandbox files back to file system
+    :param tfile: tar ball file name
+    :param jdict: JSON sandbox dictionary in form: {"path-to-json1": jdoc1, "path-to-json2": jdoc2}
+    :param dest: optional destination parameter to write final tar ball (use for unit tests)
```

Comment:
What is the problem in opening the freshly created tarball and verifying its content? That seems a reasonable solution for the unit test, and does not require any instrumentation of the actual function being tested.

Reply:
The point of writing tarball to a different location is reproduceability of unit test. I added a real tarball to test area and I want to use it over and over in unit tests. If I'll overwrite it then there is no way I can check and compare its original content with new one. In other words, yes you are right we open freshly created tar ball and this is exactly what I'm doing in unit test but I write this tarball to a different location from the one I'm reading.

Reply:
Alan, I do not know if this can be marked as resolved as I explained already that unit test should have both tarballs and writing new tarball in new place is mandatory to perform such comparison. Anyway, I'll mark this as resolved, but if you fell otherwise please explain how to deal with this in unit test which by itself I think is desired to have for such complex logic.